### ISSUE TITLE
Company-search-candiates using complete reading

### ISSUE THREAD

**Issue by jtamagnan** [company-mode/company-mode#460]
Sometimes when I am using company mode I'd like to use fuzzy matching to search through the candidates. I wrote this small function to do that, it allows the user to use whichever matching system they're used to in their minibuffer. Should I create a pull request with this feature?

```lisp
 (bind-key "M-s" 'company-search-complete-read company-active-map)
 (defun company-search-complete-read ()
   (interactive)
   (when (company-manual-begin)
     (let ((company-common-part (substring-no-properties company-common)))
       (company-finish
	(completing-read "Candidate: " company-candidates t nil company-common-part nil company-common-part nil)))))
```

**Comment by dgutov** [comment]
I'd like to keep this question open for a while. Let's see if other users are enthusiastic about it.

Regarding fuzzy matching, you might want to customize `company-search-regexp-function` instead.

**Comment by geraldus** [comment]
Hi friends!  I'm missing fuzzy search in completions list too, though I've got used to default functionality I always thinking about fuzzy search.

@dgutov there are three options for `company-search-regexp-function`, but none of them seems to correspond to fuzzy search, can you explain how should it be configured?

Thanks.

**Comment by dgutov** [comment]
The other two options are fuzzy-ish. But how about:

```el
(defun s-to-flex-regexp (s)
  (concat (regexp-quote (string (aref s 0)))
          (mapconcat (lambda (c)
                       (concat "[^" (string c) "]*"
                               (regexp-quote (string c))))
                     (substring s 1) "")))
```

**Comment by geraldus** [comment]
@dgutov , wow, fuzzy regexp is quite simple, I've never thought how can I express fuzzy match via regexp.  Thanks (:

However, your suggestion gives me error:
```elisp
(setq company-search-regexp-function #'s-to-flex-regexp)
```
> Company: frontend company-pseudo-tooltip-unless-just-one-frontend error "Args out of range: "", 0" on command post-command

Sorry, I have no time to debug right now (:

**Comment by jtamagnan** [comment]
The reason that I created my function had been because I am very used to
using ivy to do my searching and when there is a long list of candidates
sometimes that helps me. It also allows me to use the navigation hydra
built into ivy. It's not so much about fuzzy matching as it is using the
same exact searching environment that I'm used.

**Comment by dgutov** [comment]
> wow, fuzzy regexp is quite simple, I've never thought how can I express fuzzy match via regexp

It's actually the optimized version (from ido-mode, the flex matching code). Previously, we just put `.*` between all characters, but that leads to backtracking.

> However, your suggestion gives me error:

Yes, thanks. It didn't handle the empty string case. I've pushed the fixed version to the repo.

**Comment by dgutov** [comment]
@jtamagnan Ok, good to know. But you can still keep that snippet of code in your config, too.

**Comment by jtamagnan** [comment]
@dgutov Oh absolutely thats how I have it, I just thought that if it had garnered interest I would clean up the code and create a pull request for it.

**Comment by issue-dispenser** [comment]
This might be interesting, but it doesn't seem to work with helm. It doesn't give me any selections to choose from.

**Comment by jtamagnan** [comment]
Good catch! I had accidently transposed two arguments to completing-read. I've tested this code and it works in ivy and helm
``` lisp
(defun company-search-complete-read ()
    (interactive)
    (when (company-manual-begin)
      (let ((company-common-part (substring-no-properties company-common)))
	(company-finish
	 (completing-read "Candidate: " company-candidates nil t company-common-part nil company-common-part nil)))))
```

**Comment by jtamagnan** [comment]
This is a version that has some error checking in case company-common is nil. This seems to happen sometimes depending on the backend.
``` lisp
  (defun company-search-complete-read ()
    (interactive)
    (message "%S" company-common)
    (when (company-manual-begin)
      (let ((company-common-part (if (stringp company-common)
				     (substring-no-properties company-common)
				   "")))
	(company-finish
	 (completing-read "Candidate: " company-candidates nil t company-common-part nil company-common-part nil))))))
```

**Comment by dgutov** [comment]
Are there completing-read function that don't strip properties from initial-input or default?

You can use `(and company-common (substring-no-properties company-common))`. `nil` is a perfectly valid value to pass in as initial-input or default.

By the way, you should make a decision, and pass `company-common-part` as only one of the arguments. Probably the default; `initial-input` is semi-deprecated.

**Comment by jtamagnan** [comment]
I hadn't noticed that it stripped the properties beforehand. I tried in Ivy and in helm and sure enough the properties were stripped. I think that code was in there since when I was first writing it and experimenting with it.

The reason I had the string in initial-input was because I prefer when it 'primes' the minibuffer with the inital-input instead of having to press `M-n` but I guess that since initial-input is depreciated I should get used to it!

The clean version now looks like:
```
  (defun company-search-complete-read ()
    (interactive)
    (when (company-manual-begin)
      (company-finish
       (completing-read "Candidate: " company-candidates nil t nil nil company-common nil))))
```

Thank you for the help and knowledge

**Comment by dgutov** [comment]
If `company-common` is better as initial input, then it should be initial-input, regardless of semi-deprecation. It doesn't seem like it'll work well as the default value.

**Comment by andschwa** [comment]
@jtamagnan This is wonderful. Do you still use it?

**Comment by tangxinfa** [comment]
It's annoying that the automatically company popup will interrupt normal
editing. Customize it to only select company candidate with `M-<number>`, and let
the `C-n`/`C-p`/`<return>` key works as normal. When press `TAB`, search candidates with
completion frameworks(such as `live-completions`, `helm`, `ivy`, `selectrum` ...), this
works very good.

My configuration for reference purpose:

    (use-package company
      :delight
      :commands
      (my/company-complete)
      :custom
      (company-idle-delay 0.1)
      (company-show-numbers 'left) ;; M-<d> to select candidates by number
      (company-tooltip-align-annotations t)
      (company-minimum-prefix-length 2)
      (company-frontends '(company-pseudo-tooltip-frontend
                           company-echo-metadata-frontend))
      (company-dabbrev-downcase nil)
      (company-selection-default nil)
      :bind
      ("M-c" . my/company-complete)
      (:map company-active-map
            ;; Avoid interrupt normal edit
            ("C-n" . nil)
            ("C-p" . nil)
            ("RET" . my/company-bypass-ret)
            ("<return>" . my/company-bypass-ret)
            ("<C-return>" . my/company-complete-first)
            ;; Enable press TAB to select candidate
            ("TAB" . my/company-search)
            ("<tab>" . my/company-search))
      :hook (after-init . (lambda () (global-company-mode t)))
      :config
      (add-to-list 'company-transformers
                   'company-sort-by-backend-importance 'append)
      (defun my/company-complete (select-backend)
        "Trigger company complete, SELECT-BACKEND for prompt select a backend."
        (interactive "P")
        (call-interactively
         (if select-backend
             #'company-begin-backend
           #'company-manual-begin)))
      (defun my/company-search ()
        (interactive)
        (company-abort)
        (let ((company-frontends nil))
          (when (company-manual-begin)
            (company-finish
             (completing-read "Candidate: " company-candidates nil t nil nil company-common nil)))))
      (defun my/company-bypass-ret ()
        (interactive)
        (company-abort)
        (execute-kbd-macro (kbd "<RET>")))
      (defun my/company-complete-first ()
        (interactive)
        (company-select-first)
        (company-complete-selection)))

**Comment by dgutov** [comment]
@tangxinfa You can try `company-tng`, it lets you make the selection with TAB only.

It keeps the `C-n`/`C-p` bindings, but you can unset them through `company-tng-map`.

**Comment by tangxinfa** [comment]
@dgutov

Thanks, `company-tng` has some help here, but i prefer to call
`company-complete-common-or-cycle` instead when press `TAB`, for avoid company
popup interrupt normal editing and keep company operations works as usual, i will do
the following:

1, `(setq company-selection-default nil)`

Avoid select candidate automatically, so `C-n` `C-p` `RET` can do normal editing.

2, `(setq company-require-match nil)`

Avoid force select candidate, so `C-n` `C-p` `RET` will can normal editing.

3, rebinding `company-active-map`

    "C-n"          company-select-next if has selected candidate otherwise move to next line
    "C-p"          company-select-next if has selected candidate otherwise move to next line
    "<return>/RET" company-complete-selection if has selected candidate otherwise insert newline
    "<tab>/TAB"    company-complete-common-or-cycle
